LUCENE-9334 Consistency of field data structures

mayya-sharipova opened the pull request:
Require consistency between data-structures on a per-field basis

A field must be indexed with the same index options and data-structures across
all documents within a segment. Thus, for example, it is not allowed to have
one document in a segment where a certain field is indexed with doc values
and points, and another document where the same field is indexed only with
points. But it is allowed for a document not to have a certain field at all.

This requirement is ensured only per segment basis, thus it is possible
to have a index where two segments have different data structures
for the same field.

This patch also corrects tests where fields were indexed inconsistently
with different data structures across documents:
- Remove TestDocValuesIndexing::testDocsWithField, as it doesn't
test anything special, what it was originally created to test

jpountz commented:
> This requirement is ensured only per segment basis, thus it is possible
to have a index where two segments have different data structures
for the same field.

Could we make this work across segments?

mayya-sharipova commented:
> Could we make this work across segments?

@jpountz  Thanks for taking a look. I just wanted to confirm that we are ok with this change, which means for example, that we can't enable  doc_values on a field where doc values were previously disabled.

jpountz commented:
@mayya-sharipova Yes, if the field has been added previously without doc values, then it should be illegal to add it later with doc values indeed.

jpountz commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -130,127 +167,255 @@ public boolean checkConsistency() {
+              + fieldName
               + "\" from points dimensionCount="
-              + this.pointDimensionCount
+              + dc1
```
I liked the descriptive variable names better.

mayya-sharipova replied:
Addressed in 6e7540ebd0ef79536cffabcf0ddc7a592b792252

jpountz commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -130,127 +167,255 @@ public boolean checkConsistency() {
+  }
+
+  /**
+   * Very that the provided posting options are the same
```
Several javadocs seem to have this typo.

```suggestion
   * Verify that the provided posting options are the same
```

mayya-sharipova replied:
Addressed in 6e7540ebd0ef79536cffabcf0ddc7a592b792252

jpountz commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -1313,4 +1259,110 @@ public void recycleIntBlocks(int[][] blocks, int offset, int length) {
     }
   }
+
+  private static final class FieldSchema {
```
can you add a bit more docs, e.g. how this it differ from FieldInfo?

mayya-sharipova replied:
Addressed in 6e7540ebd0ef79536cffabcf0ddc7a592b792252

jpountz reviewed (commented):
Hi Mayya, I only had a quick look at the PR for now as the diff is pretty large. Can you give a bit more details about how this PR works at a high level? E.g. how does it handle the case when two threads are concurrently trying to add a field with different schemas?

mayya-sharipova commented:
@jpountz  Adrien, thank a lot for the initial review. Sorry for the delay, the patch turned out to be much more involved than I expected.
There are still few tests failing, but I will try to fix them soon.


>  Can you give a bit more details about how this PR works at a high level? E.g. how does it handle the case when two threads are concurrently trying to add a field with different schemas?

---
**Segment level**: ensuring a field has the same schema across all the documents of the segment.
- We use `FieldSchema` for this, that's an internal field of `PerField`. It represents a schema of the field in the current document. 
- With every new document we reset `FieldSchema`.  As the document fields are processed, we update the schema with options encountered in this document. Once the processing for the document is done, we compare the built `FieldSchema` of the current document with the corresponding `FieldInfo `(FieldInfo is built on a first document in which we encounter this field).

Relevant code in `IndexingChaing:processDocument`:
```java
...
       if (pf.fieldGen != fieldGen) { // first time we see this field in this document
          ....
          pf.reset(docID);
        }
...
for (int i = 0; i < fieldCount; i++) {
  PerField pf = fields[i];
  if (pf.fieldInfo == null) { // the first time we see this field in this segment
    initializeFieldInfo(pf);
  } else {
    pf.schema.assertSameSchema(pf.fieldInfo);
  }
}
```

---
**Index level**: ensuring a field has the same schema across all the documents of the index.

- This check is done in `FieldInfos`. 
- When we encounter a new field in a segment, we try to initialize it in `IndexingChain::initializeFieldInfo` with  the options from the current `FieldSchema`.  This calls `FieldInfos.Builder::add` ->  `FieldInfos.Builder::addField`.  
-  The  first thing in `FieldInfos.Builder::addField`  is to call `globalFieldNumbers.addOrGet` with the given schema options. `globalFieldNumbers.addOrGet`  is a synchronized method, and as I understood  `FieldNumbers` are shared across all indexing threads of the same index.  
- `globalFieldNumbers.addOrGet` for a field it sees the first time, will initialize all its maps `indexOptions`, `docValuesType` etc.  If the field already exists, it will check that the given in parameters schema options are the same as stored in its maps for the given field.
- As `globalFieldNumbers.addOrGet` is a synchronized method only a single thread will be able to initialize schema options for a field. All other threads that deal with the same field, must confirm with the same field schema.

@jpountz  Is my assumption and logic correct here?

jpountz commented:
I need to spend more time understanding how everything ties together but the logic you described makes sense to me.

mayya-sharipova commented:
@jpountz  Adrien, I also wanted to ask your opinion about the behaviour of `writer.updateDocValues(Term term, fieldX)`:
1.  What should happen if `fieldX` doesn't exist? 
2. What should happen if `fieldX` doesn't have docValues enabled?

Before in both of these cases `updateDocValues` worked well creating a field if it doesn't exist or enabling doc values.

But with current PR as the field schema is fixed, we can:
1. Allow the 1st case  with the assumption that the created field will only has docValues and nothing more.
2. Throw an error on 2nd case, as docValues are not enabled.

WDYT?

jpountz commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -349,7 +350,8 @@ private void writePoints(SegmentWriteState state, Sorter.DocMap sortMap) throws
             perField.pointValuesWriter.flush(state, sortMap, pointsWriter);
             perField.pointValuesWriter = null;
-          } else if (perField.fieldInfo.getPointDimensionCount() != 0) {
+          } else if (perField.fieldInfo != null
```
shouldn't fieldInfo always be non null when reaching here?

mayya-sharipova replied:
Not, always.
We first create PerField: `PerField pf = getOrAddPerField(field.name(), fieldType);` and then initialize its FieldInfo `initializeFieldInfo(pf)`. Initialization of FieldInfo may fail, if there is already a global field with the same name but a different schema. So we may end up with `PerField` with a null `fieldInfo`.

jpountz commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -399,7 +401,8 @@ private void writeDocValues(SegmentWriteState state, Sorter.DocMap sortMap) thro
             perField.docValuesWriter.flush(state, sortMap, dvConsumer);
             perField.docValuesWriter = null;
-          } else if (perField.fieldInfo.getDocValuesType() != DocValuesType.NONE) {
+          } else if (perField.fieldInfo != null
```
shouldn't fieldInfo always be non null when reaching here?

mayya-sharipova replied:
Not, always.
We first create PerField: `PerField pf = getOrAddPerField(field.name(), fieldType);` and then initialize its FieldInfo `initializeFieldInfo(pf)`. Initialization of FieldInfo may fail, if there is already a global field with the same name but a different schema. So we may end up with `PerField` with a null `fieldInfo`.

jpountz reviewed (commented):
The logic makes sense to me. The PR is massive and this logic is tricky so I haven't had time to go through all the details yet but it looks correct to me. One aspect I wonder about is the impact on indexing times, have you had a chance to run indexing benchmarks?

mayya-sharipova commented:
@jpountz  Hi Adrien. Thanks for your review of the PR: https://github.com/apache/lucene-solr/pull/2186. I will go through your review and address your comments. I wanted to ask two assumptions I have, and check if we are ok with them: 

1) A first doc with a field introduces FieldInfo for this field for the whole index, even if eventually this doc doesn't get indexed (e.g. if this doc has a too big stored field, the whole doc will be rolled back and deleted).  

2) Doc values updates (`IndexWriter#updateDocValues`) are only applicable for fields that are indexed with doc values only. This was not the case before, and for example we could update doc values for a field that was indexed with postings.

jpountz commented:
@mayya-sharipova These assumptions sound right to me.

jpountz commented on lucene/test-framework/src/java/org/apache/lucene/index/BasePointsFormatTestCase.java:
```diff
@@ -1174,8 +1174,11 @@ public void testMixedSchema() throws Exception {
+        expectThrows(IllegalArgumentException.class, () -> w.addDocument(doc));
+    assertEquals(
+        "cannot change field \"id\" from index options=DOCS to inconsistent index options=NONE",
+        ex.getMessage());
```
I think we should refactor or drop this test, as it is not testing the points format now, but IndexingChain/FieldsInfos' logic. Maybe we could rename the test `testMergeMissing` and configure the first segment to not have the `id` field at all.

mayya-sharipova replied:
The test refactored [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-974afcef788664579ec220b758dd40bf67d8e72f3fb214e52016b8465fb54289L1161)

mayya-sharipova commented:
@jpountz  Thanks for your feedback. I have [moved](https://github.com/apache/lucene/pull/11)  the PR to the new Lucene repo and will address your feedback there.

Closing this PR.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -111,6 +111,43 @@ public FieldInfo(
 
   /** Performs internal consistency checks. Always returns true (or throws IllegalStateException) */
   public boolean checkConsistency() {
+    return checkOptionsCorrectness(
```
addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b9ceb6f65b3597b6c2406c2d41cc7db39e04e21bb77a13fcc92eb0429c444d46L135)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -130,127 +167,252 @@ public boolean checkConsistency() {
-                + " to inconsistent index options="
-                + indexOptions);
-      }
+  void verifySameSchema(FieldInfo o, long dvGen) {
```
addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b9ceb6f65b3597b6c2406c2d41cc7db39e04e21bb77a13fcc92eb0429c444d46L244)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -130,127 +167,252 @@ public boolean checkConsistency() {
+   *
+   * @throws IllegalArgumentException if they are not the same
+   */
+  public static void verifySameIndexOptions(
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b9ceb6f65b3597b6c2406c2d41cc7db39e04e21bb77a13fcc92eb0429c444d46L274)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfo.java:
```diff
@@ -130,127 +167,252 @@ public boolean checkConsistency() {
+      verifySameStoreTermVectors(fieldName, this.storeTermVector, o.storeTermVector);
+    }
+    verifySameDocValuesType(fieldName, this.docValuesType, o.docValuesType);
+    verifySameDVGen(fieldName, this.dvGen, dvGen);
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b9ceb6f65b3597b6c2406c2d41cc7db39e04e21bb77a13fcc92eb0429c444d46L252)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -455,8 +373,42 @@ synchronized int addOrGet(
         assert fieldNumber >= 0;
+        FieldInfo.checkOptionsCorrectness(
+            fieldName,
+            false,
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0L378)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -481,205 +433,102 @@ synchronized int addOrGet(
-      }
+    /**
+     * Returns true if the {@code fieldName} exists in the map and is of the same {@code dvType},
+     * and it is docValues only field.
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0L479)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -972,6 +671,64 @@ private boolean assertNotFinished() {
       return true;
     }
 
+    private FieldInfo addField(
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0L674)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -1860,8 +1862,8 @@ public long softUpdateDocument(
-      throw new IllegalArgumentException("can only update existing numeric-docvalues fields!");
+    if (globalFieldNumberMap.containsDvOnlyField(field, DocValuesType.NUMERIC) == false) {
+      throw new IllegalArgumentException("can only update existing numeric docvalues only fields!");
     }
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0L482)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -79,6 +79,7 @@
 
   // Holds fields seen in each document
   private PerField[] fields = new PerField[1];
+  private PerField[] docFields = new PerField[10];
```
@jpountz  I wonder what should be a smaller initial size for the number of fields in a doc?

I changed the size to 2 [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-1e11574161aacb332207eea2eb36edbf86bf3518416e4c089561f3c0fbe43e03L82)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -581,37 +584,59 @@ private void finishStoredFields() throws IOException {
-    int fieldCount = 0;
-
+    int fieldCount =
+        0; // How many field names we've seen (collapses multiple field instances by the same name)
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-1e11574161aacb332207eea2eb36edbf86bf3518416e4c089561f3c0fbe43e03L587)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/IndexingChain.java:
```diff
@@ -581,37 +584,59 @@ private void finishStoredFields() throws IOException {
-    // instances of a given field at once, since the
-    // analyzer is free to reuse TokenStream across fields
-    // (i.e., we cannot have more than one TokenStream
-    // running "at once"):
```
I am not sure, as I did not study deeply of termHash. But just in case, I've added this comment back [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-1e11574161aacb332207eea2eb36edbf86bf3518416e4c089561f3c0fbe43e03R593)

mayya-sharipova commented on lucene/test-framework/src/java/org/apache/lucene/index/BaseVectorFormatTestCase.java:
```diff
@@ -131,7 +115,8 @@ public void testIllegalDimChangeTwoDocs() throws Exception {
       w.addDocument(doc);
-      if (random().nextBoolean()) {
+      boolean rb = random().nextBoolean();
+      if (rb) {
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-0761a46b20367b4f8c4c85e370dd0156270193e5b897064369cfae68523e81bdL118)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -481,205 +433,102 @@ synchronized int addOrGet(
     /**
-     * Returns true if the {@code fieldName} exists in the map and is of the same {@code dvType}.
+     * Construct a new FieldInfo based on the options in global field numbers This method is not
+     * synchronized as all the options it uses are not modifiable.
```
From the point of view `FieldInfos.FieldNumbers`,  doc values type is not modifiable and set only once on adding this field to global field numbers.

This method `constructFieldInfo` also checks the the provided in arguments `dvType` matches the one stored in the global field numbers for this field.  So I am not sure how it can be modifiable.

Please let me know if I am missing something.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -481,205 +433,102 @@ synchronized int addOrGet(
+
+      boolean isSoftDeletesField = fieldName.equals(softDeletesFieldName);
+      FieldDimensions dims = dimensions.get(fieldName);
+      FieldVectorProperties vectors = vectorProps.get(fieldName);
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0L504)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -796,167 +565,97 @@ public void add(FieldInfos other) {
     }
 
-    private FieldInfo addOrUpdateInternal(
+    public FieldInfo add(
```
Addressed [here](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5#diff-b81932cf39b70347481b9659dd55b9b373427c2992cdacdfbc10ff8f4ea9eec0R745)

mayya-sharipova commented:
@jpountz Thanks a lot for the second round of review. I've addressed these comments in the [new repo](https://github.com/apache/lucene/pull/11/commits/0fe3493110ac2a5f750ad41f732436daff6c69f5).

I've resolved all comments that I fully addressed, and left comments on some that need more discussion.
